pkg/requester/selection/any.go

```diff
@@ -0,0 +1,129 @@
+	return &anyNodeSelector{
+		nodeDiscoverer:       params.NodeDiscoverer,
+		nodeRanker:           params.NodeRanker,
+		overAskForBidsFactor: system.Max(1, params.OverAskForBidsFactor),
+	}
+}
+
+func (s *anyNodeSelector) Select(ctx context.Context, job *model.Job, minCount, desiredCount int) ([]model.NodeInfo, error) {
+	ctx = log.Ctx(ctx).With().Str("JobID", job.ID()).Logger().WithContext(ctx)
+
+	nodeIDs, err := s.nodeDiscoverer.FindNodes(ctx, *job)
+	if err != nil {
+		return nil, err
+	}
+	log.Ctx(ctx).Debug().Int("Discovered", len(nodeIDs)).Msg("Found nodes for job")
+
+	rankedNodes, err := s.nodeRanker.RankNodes(ctx, *job, nodeIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	// filter nodes with rank below 0
+	var filteredNodes []requester.NodeRank
+	for _, node := range rankedNodes {
+		if node.Rank >= 0 {
+			filteredNodes = append(filteredNodes, node)
+		}
+	}
+	rankedNodes = filteredNodes
+	log.Ctx(ctx).Debug().Int("Ranked", len(rankedNodes)).Msg("Ranked nodes for job")
+
+	if len(rankedNodes) < minCount {
+		err = requester.NewErrNotEnoughNodes(minCount, len(rankedNodes))
+		return nil, err
+	}
+
+	sort.Slice(rankedNodes, func(i, j int) bool {
+		return rankedNodes[i].Rank > rankedNodes[j].Rank
+	})
+
+	selectedNodes := rankedNodes[:system.Min(len(rankedNodes), desiredCount)]
+	selectedInfos := generic.Map(selectedNodes, func(nr requester.NodeRank) model.NodeInfo { return nr.NodeInfo })
+	return selectedInfos, nil
+}
+
+func (s *anyNodeSelector) SelectNodes(ctx context.Context, job *model.Job) ([]model.NodeInfo, error) {
+	minCount := system.Max(job.Spec.Deal.MinBids, job.Spec.Deal.Concurrency)
+	desiredCount := minCount * int(s.overAskForBidsFactor)
+	return s.Select(ctx, job, minCount, desiredCount)
+}
+
+func (s *anyNodeSelector) SelectNodesForRetry(ctx context.Context, job *model.Job, jobState *model.JobState) ([]model.NodeInfo, error) {
+	// calculate how many executions we still need, and evaluate if we can ask more nodes to bid
+	var minExecutions int
+	if job.Spec.Deal.MinBids > 0 && jobState.ReceivedBidCount() < job.Spec.Deal.MinBids {
+		// if we are still queuing bids, then we need at least MinBids to start accepting bids
+		minExecutions = system.Max(job.Spec.Deal.GetConcurrency(), job.Spec.Deal.MinBids)
+	} else if jobState.PublishedOrPublishingCount() > 0 {
+		// if at least a single execution was published or still publishing, then we don't need to retry in case some executions failed to publish
+		minExecutions = 1
+	} else {
+		// by default, we need executions as many as the job concurrency
+		minExecutions = job.Spec.Deal.GetConcurrency()
+	}
+
+	enoughExecutions := jobState.NonDiscardedCount() >= minExecutions
+	if enoughExecutions {
+		// There are enough running executions, so no action to take right now.
+		return []model.NodeInfo{}, nil
+	}
+
+	desiredNodeCount := minExecutions - jobState.NonDiscardedCount()
+	return s.Select(ctx, job, desiredNodeCount, desiredNodeCount)
```

Comment:
Similar to my comment in the "all selector", do we need to re-discover nodes here? (`Select` will rerun node discovery)

Reply:
In this case it's useful, because between the job being orchestrated and this moment, new nodes that might be able to run the job may have joined the network. So it makes sense to re-orcehstrate this execution with as current a view as possible of what nodes are available.